src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,485 @@
+    });
+
+    // Update the aria-described by when the number of errors changes.
+    this._errorChildren.changes.subscribe(() => {
+      this._syncDescribedByIds();
+      this._changeDetectorRef.markForCheck();
+    });
+
+    // Initial mat-hint validation and subscript describedByIds sync.
+    this._validateHints();
+    this._syncDescribedByIds();
+  }
+
+  /** Throws an error if the form field's control is missing. */
+  private _assertFormFieldControl() {
+    if (!this._control) {
+      throw getMatFormFieldMissingControlError();
+    }
+  }
+  private _updateFocusState() {
+    // Usually the MDC foundation would call "activateFocus" and "deactivateFocus" whenever
+    // certain DOM events are emitted. This is not possible in our implementation of the
+    // form-field because we support abstract form field controls which are not necessarily
+    // of type input, nor do we have a reference to a native form-field control element. Instead
+    // we handle the focus by checking if the abstract form-field control focused state changes.
+    if (this._control.focused && !this._isFocused) {
+      this._isFocused = true;
+      this._foundation.activateFocus();
+    } else if (!this._control.focused && (this._isFocused || this._isFocused === null)) {
+      this._isFocused = false;
+      this._foundation.deactivateFocus();
+    }
+  }
+
+  _rerenderOutlineNotch() {
+    if (this._floatingLabel && this._hasOutline()) {
+      this._foundation.notchOutline(this._shouldLabelFloat());
+    }
+  }
+
+  /** Whether the floating label should always float or not. */
+  _shouldAlwaysFloat() {
+    return this.floatLabel === 'always';
+  }
+
+  _hasOutline() {
+    return this.appearance === 'outline';
+  }
+
+  _hasFloatingLabel() {
+    return !!this._labelChildNonStatic || !!this._labelChildStatic;
+  }
+
+  _shouldLabelFloat() {
+    return this._control.shouldLabelFloat || this._shouldAlwaysFloat();
+  }
+
+  _isStandardAppearance() {
+    return this.appearance === 'standard' || this.appearance === 'legacy';
+  }
+
+  /** Determines whether a class from the NgControl should be forwarded to the host element. */
+  _shouldForward(prop: keyof NgControl): boolean {
+    const ngControl = this._control ? this._control.ngControl : null;
+    return ngControl && ngControl[prop];
+  }
+
+  /** Determines whether to display hints or errors. */
+  _getDisplayedMessages(): 'error' | 'hint' {
+    return (this._errorChildren && this._errorChildren.length > 0 &&
+      this._control.errorState) ? 'error' : 'hint';
+  }
+
+  /** Does any extra processing that is required when handling the hints. */
+  private _processHints() {
+    this._validateHints();
+    this._syncDescribedByIds();
+  }
+
+  /**
+   * Ensure that there is a maximum of one of each `<mat-hint>` alignment specified, with the
+   * "hintLabel" attribute being considered as `align="start"`.
+   */
+  private _validateHints() {
+    if (this._hintChildren) {
```

Comment:
We could probably make this check only happen in dev mode

Reply:
yeah. I took this from the existing form-field. Not sure why it's not done there, but that sounds good to me.